src/main/java/cn/nukkit/entity/passive/EntityVillager.java

```diff
@@ -385,4 +405,90 @@ public boolean onInteract(Player player, Item item, Vector3 clickedPos) {
     public TradeInventory getInventory() {
         return inventory;
     }
+
+    @PowerNukkitXOnly
+    @Since("1.20.0-r2")
+    public int getTradeSeed() {
+        return tradeSeed;
+    }
+
+    @PowerNukkitXOnly
+    @Since("1.20.0-r2")
+    protected void setTradeSeed(int tradeSeed) {
+        this.tradeSeed = tradeSeed;
+        this.namedTag.putInt("tradeSeed", tradeSeed);
+    }
+
+
+    public void addExperience(int xp) {
+
+        this.tradeExp += xp;
+        this.setDataProperty(new IntEntityData(DATA_TRADE_EXPERIENCE, this.tradeExp));
+        int next = getTradeTier()+1;
+        if(next < this.tierExpRequirement.length) {
+            if(tradeExp >= this.tierExpRequirement[next]) {
+                setTradeTier(next+1);
+            }
+        }
+
+    }
+
+    @Since("1.20.0-r2")
+    @Override
+    public boolean onUpdate(int tick) {
+
+        if(tick%100 == 0) {
+
+            if(profession != 0) {
+                if(recipes.getAll().size() == 0) applyProfession(Profession.getProfession(this.profession));
+            }
+
+            if(tradeExp == 0 && !this.namedTag.contains("traded")) {
+                boolean professionFound = false;
+                for(int x = -1; x <= 1; x++) {
+                    for (int z = -1; z <= 1; z++) {
+                        Block block = getLocation().add(x, 0, z).getLevelBlock();
+                        int id = block.getId();
+                        for (Profession profession : Profession.getProfessions().values()) {
+                            if (id == profession.getBlockid()) {
+                                professionFound = true;
+                                if (this.profession != profession.getIndex()) {
+                                    this.setTradeSeed(new Random().nextInt(Integer.MAX_VALUE));
```

Comment:
Using ThreadLocalRandom or NukkitRandom from the level of this villager instead, Random is globally synchronized and may block PNX, thereby affecting performance.